receiver/ocmetrics/opencensus.go

```diff
@@ -0,0 +1,162 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ocmetrics
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"google.golang.org/api/support/bundler"
+
+	"go.opencensus.io/trace"
+
+	commonpb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	agentmetricspb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/metrics/v1"
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	"github.com/census-instrumentation/opencensus-service/internal"
+	"github.com/census-instrumentation/opencensus-service/metricsink"
+)
+
+// Receiver is the type used to handle metrics from OpenCensus exporters.
+type Receiver struct {
+	metricSink         metricsink.Sink
+	metricBufferPeriod time.Duration
+	metricBufferCount  int
+}
+
+// New creates a new opencensus.Receiver reference.
+func New(sr metricsink.Sink, opts ...Option) (*Receiver, error) {
+	if sr == nil {
+		return nil, errors.New("needs a non-nil metricsink.Sink")
+	}
+	ocr := &Receiver{metricSink: sr}
+	for _, opt := range opts {
+		opt.WithReceiver(ocr)
+	}
+	return ocr, nil
+}
+
+var _ agentmetricspb.MetricsServiceServer = (*Receiver)(nil)
+
+type bundledMetrics struct {
+	metrics  []*metricspb.Metric
+	node     *commonpb.Node
+	resource *resourcepb.Resource
+}
+
+var errMetricsExportProtocolViolation = errors.New("protocol violation: Export's first message must have a Node")
+
+const receiverName = "opencensus_metrics"
+
+// Export is the gRPC method that receives streamed metrics from
+// OpenCensus-metricproto compatible libraries/applications.
+func (ocr *Receiver) Export(mes agentmetricspb.MetricsService_ExportServer) error {
+	// The bundler will receive batches of metrics i.e. []*metricspb.Metric
+	// We need to ensure that it propagates the receiver name as a tag
+	ctxWithReceiverName := internal.ContextWithReceiverName(mes.Context(), receiverName)
+	metricsBundler := bundler.NewBundler((*bundledMetrics)(nil), func(payload interface{}) {
+		ocr.batchMetricExporting(ctxWithReceiverName, payload)
+	})
+
+	metricBufferPeriod := ocr.metricBufferPeriod
+	if metricBufferPeriod <= 0 {
+		metricBufferPeriod = 2 * time.Second // Arbitrary value
+	}
+	metricBufferCount := ocr.metricBufferCount
+	if metricBufferCount <= 0 {
+		// TODO: (@odeke-em) provide an option to disable any buffering
+		metricBufferCount = 50 // Arbitrary value
+	}
+
+	metricsBundler.DelayThreshold = metricBufferPeriod
+	metricsBundler.BundleCountThreshold = metricBufferCount
+
+	// The first message MUST have a non-nil Node.
```

Comment:
nit: the comment doesn't match the check below.

Reply:
Updated the comment a bit.